src/controller/src/lib.rs

```diff
@@ -142,6 +142,16 @@ impl<T> Controller<T> {
     pub fn active_compute(&mut self) -> ActiveComputeController<T> {
         self.compute.activate(&mut *self.storage)
     }
+
+    /// Remove orphaned services from the orchestrator.
+    pub async fn remove_orphans(
+        &mut self,
+        next_ids: (GlobalId, ReplicaId),
+    ) -> Result<(), anyhow::Error> {
+        self.storage.remove_orphans(next_ids.0).await?;
+        self.compute.remove_orphans(next_ids.1).await?;
```

Comment:
For orphaned indexes (or dataflows), the rehydration logic should take care of it. In case of a crash, there is no orphans as the computeds are stateless and wait to be rehydrated from environmentd.

Entire compute clusters don't have that problem either, because the only externally persisted state are the replica pod entries in kubernetes. So we should be good there. 

I do have another potential source of inconsistencies in mind: The builtin table updates that don't run transactionally with the catalog update. (<- let me know if this is not true). So if envd crashes right after a drop replica has been communicated to the catalog, are we sure the builtin table updates are always correct?

If you know of any other state that envd modifies (outside the catalog) let me know!

Reply:
Builtin tables should be fine. As part of bootstrapping, the Coord identifies what all the system tables should look like and what they currently look like, and will send the needed appends to get them into the correct state. 

https://github.com/MaterializeInc/materialize/blob/796cd14c2132b95ef22244b5c0ca6fee23744c85/src/adapter/src/coord.rs#L865-L892